File: mm-php/src/io/metamessage/mm/CamelToSnake.php
<?php

namespace io\metamessage\mm;

class CamelToSnake {
    public static function convert(string $s): string {
        if (empty($s)) {
            return '';
        }
        $result = '';
        for ($i = 0; $i < strlen($s); $i++) {
            $c = $s[$i];
            if (ctype_upper($c)) {
                if ($i > 0) {
                    $prev = $s[$i - 1];
                    $prevUpper = ctype_upper($prev);
                    $nextUpper = $i + 1 < strlen($s) && ctype_upper($s[$i + 1]);
                    if (!$prevUpper || ($i + 1 < strlen($s) && !$nextUpper)) {
                        $result .= '_';
                    }
                }
                $result .= strtolower($c);
            } else {
                $result .= $c;
            }
        }
        return $result;
    }
}

File: mm-java/src/main/java/io/metamessage/mm/CamelToSnake.java
package io.github.metamessage.mm;

public final class CamelToSnake {
    private CamelToSnake() {}

    public static String convert(String s) {
        if (s == null || s.isEmpty()) {
            return "";
        }
        StringBuilder result = new StringBuilder(s.length() + 4);
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (Character.isUpperCase(c)) {
                if (i > 0) {
                    char prev = s.charAt(i - 1);
                    boolean prevUpper = Character.isUpperCase(prev);
                    boolean nextUpper = i + 1 < s.length() && Character.isUpperCase(s.charAt(i + 1));
                    if (!prevUpper || (i + 1 < s.length() && !nextUpper)) {
                        result.append('_');
                    }
                }
                result.append(Character.toLowerCase(c));
            } else {
                result.append(c);
            }
        }
        return result.toString();
    }
}
